### PR TITLE
lib/db: Fix unlikely race

### PR DESCRIPTION
**Pull request by imsodin**
This one is quite ridiculous, nevertheless it happened when repeatedly running this test (for entirely different reasons):

```
=== RUN   TestRequestRemoteRenameChanged
23:14:39 INFO: No stored folder metadata for "default": recalculating
23:14:39 INFO: Ready to synchronize "default" (default) (sendreceive)
23:14:39 INFO: Completed initial scan of sendreceive folder "default" (default)
23:14:39 INFO: Device AIR6LPZ-7K4PTTV-UXQSMUU-CPQ5YWH-OEDFIIQ-JUG777G-2YQXXR5-YD6AWQR client is " " named "" at 
==================
WARNING: DATA RACE
Write at 0x00c001fd26fd by goroutine 63:
  runtime.slicecopy()
      /media/ext4_data/Linux/source/go/src/runtime/slice.go:197 +0x0
  github.com/syncthing/syncthing/lib/db.(*NamespacedKV).Delete()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/db/namespaced.go:157 +0x158
  github.com/syncthing/syncthing/lib/fs.(*MtimeFS).save()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/fs/mtimefs.go:97 +0x12a
  github.com/syncthing/syncthing/lib/fs.(*MtimeFS).Chtimes()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/fs/mtimefs.go:65 +0x1cf
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).performFinish()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:1539 +0x3b4
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).finisherRoutine()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:1554 +0x11b7
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullerIteration.func4()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:267 +0x60

Previous write at 0x00c001fd26fd by goroutine 212:
  runtime.slicecopy()
      /media/ext4_data/Linux/source/go/src/runtime/slice.go:197 +0x0
  github.com/syncthing/syncthing/lib/db.NamespacedKV.Bytes()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/db/namespaced.go:124 +0xf7
  github.com/syncthing/syncthing/lib/db.(*NamespacedKV).Bytes()
      <autogenerated>:1 +0xaf
  github.com/syncthing/syncthing/lib/fs.(*MtimeFS).load()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/fs/mtimefs.go:114 +0xc9
  github.com/syncthing/syncthing/lib/fs.(*MtimeFS).Lstat()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/fs/mtimefs.go:75 +0xd7
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).checkParent()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:653 +0x12a
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).processNeeded()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:457 +0x866
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullerIteration()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:271 +0x763
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pull()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:198 +0xa10
  github.com/syncthing/syncthing/lib/model.(*folder).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder.go:138 +0x102f
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).Serve()
      <autogenerated>:1 +0x43
  github.com/thejerf/suture.(*Supervisor).runService.func1()
      /media/ext4_data/Coding/go/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go:600 +0x6b

Goroutine 63 (running) created at:
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullerIteration()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:266 +0x704
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pull()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:198 +0xa10
  github.com/syncthing/syncthing/lib/model.(*folder).Serve()
      /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/folder.go:138 +0x102f
  github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).Serve()
      <autogenerated>:1 +0x43
  github.com/thejerf/suture.(*Supervisor).runService.func1()
      /media/ext4_data/Coding/go/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go:600 +0x6b

Goroutine 212 (running) created at:
  github.com/thejerf/suture.(*Supervisor).runService()
      /media/ext4_data/Coding/go/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go:588 +0x68
  github.com/thejerf/suture.(*Supervisor).Serve()
      /media/ext4_data/Coding/go/pkg/mod/github.com/thejerf/suture@v3.0.2+incompatible/supervisor.go:476 +0x1033
==================
23:14:39 INFO: Puller (folder "default" (default), file "b"): finisher: file modified but not rescanned; will try again later
--- FAIL: TestRequestRemoteRenameChanged (0.28s)
    testing.go:806: race detected during execution of test
```

### CODE REVIEW

**Comment by calmh** [comment]
I admit I don't understand the race, can you explain it? It looks like it means the write is on the copy destination, but it can't be because that should be a new allocation on each run of append(). Unless the original `prefix: []byte(someString)` returns a slice with a cap longer than the length - I didn't expect that to happen but I guess it's also maybe not guaranteed according to the spec (I didn't check) - is that what's happening?

File: lib/db/namespaced.go [hunk 1]
```diff
@@ -17,16 +17,18 @@ import (
...
 
 // NewNamespacedKV returns a new NamespacedKV that lives in the namespace
 // specified by the prefix.
 func NewNamespacedKV(db *Lowlevel, prefix string) *NamespacedKV {
 	return &NamespacedKV{
-		db:     db,
-		prefix: []byte(prefix),
+		db:        db,
+		prefix:    []byte(prefix),
+		prefixLen: len(prefix),
```
Comment (calmh) [inline lib/db/namespaced.go]: If this is just an optimization I think we don't need it. `len(prefix)` is already just a read of the length word in the slice header, even though it looks like a function call.

**Comment by calmh** [comment]
Right, that is indeed what happens: https://play.golang.org/p/gXRYYxJGS7O

Given that, can I suggest that we instead set the cap explicitly in the constructor and still use the append idiom in your new prefixedKey function? I think that would be clearer and avoid someone trying to introduce the same bug again in an optimization?

```
prefixBs := []byte(prefix)
// After the conversion from string the cap will be larger than the len (in Go 1.11.5,
// 32 bytes cap for small strings). We need to cut it down to ensure append() calls
// on the prefix make a new allocation.
prefixBs = prefixBs[:len(prefixBs):len(prefixBs)]
return &NamespacedKV{
...
```

**Comment by imsodin** [comment]
Good find, even more so as with go1.12 the capacity is equal to the length of the string. I must admit I just knew that append does not necessarily allocate and didn't even think about how this might have happened.

**Pull request merged**